lib/mpool/lib/mclass.c

```diff
@@ -38,10 +41,59 @@ struct media_class {
     enum mclass_id      mcid;
     bool                gclose;
     bool                directio;
+    uint16_t            ra_pages;
     char *              dpath;
     char *              upath;
 };
 
+static merr_t
+get_read_ahead_pages(DIR *dirp, uint16_t *ra_pages)
+{
+    struct stat sbuf;
+    unsigned int major, minor;
+    char buf[PATH_MAX];
+    char line[16];
+    int fd, rc, n;
+    FILE *fp;
+
+    fd = dirfd(dirp);
+    if (fd == -1)
+        return merr(errno);
+
+    rc = fstat(fd, &sbuf);
+    if (rc == -1)
+        return merr(errno);
+
+    *ra_pages = MPOOL_RA_PAGES_DFLT;
+
+    major = major(sbuf.st_dev);
+    minor = minor(sbuf.st_dev);
+
+    n = snprintf(buf, sizeof(buf), "/sys/dev/block/%u:%u/queue/read_ahead_kb", major, minor);
+    if (ev(n < 0 || n >= sizeof(buf)))
+        return 0;
+
+    fp = fopen(buf, "r");
+    if (ev(!fp))
+        return 0;
+
+    if (fgets(line, sizeof(line), fp)) {
+        uint32_t val;
+
+        n = sscanf(line, "%u", &val);
+        if (n == 1) {
+            val >>= (ilog2(PAGE_SIZE) - 10);
+            *ra_pages = clamp_t(uint32_t, val, MPOOL_RA_PAGES_MIN, MPOOL_RA_PAGES_MAX);
+        } else {
+            ev(1);
+        }
+    }
+
+    fclose(fp);
+
+    return 0;
```

Comment:
Use hse_readfile() here to replace fopen/fgets.  That way if/when Linux gets a readfile() call we can fix it in one place.  I would prob use strtoul() here rather than sscanf...

Reply:
Changed the code to use hse_readfile(), didn't know we got an interface in platform for this :-)